keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,519 @@
+default behavior is insufficient.
+
+## Motivation
+
+Kubernetes has outgrown the [SSH tunnels](https://github.com/kubernetes/kubernetes/issues/54076).
+They complicate KAS code and only one cloud provider implemented them.
+After a year of deprecation time, they will be removed in an upcoming release.
+
+In retrospect, having an explicit level of indirection that separates user-initiated network traffic from API
+server-initiated traffic is a useful concept.
+Cloud providers want to control how API server to pod, node and service network traffic is implemented.
+Cloud providers may choose to run their API server (control network) and the cluster nodes (cluster network)
+on isolated networks. The control and cluster networks may have overlapping IP addresses.
+There for they require a non IP routing layer (SSH tunnel are an example).
+Adding this layer enables metadata audit logging. It allows validation of outgoing API server connections.
+Structuring the API server in this way is a forcing function for keeping architectural layering violations out of apiserver.
+In combination with a firewall, this separation of networks protects against security concerns such as
+[Security Impact of Kubernetes API server external IP address proxying](https://groups.google.com/d/msg/kubernetes-security-announce/tyd-MVR-tY4/tyREP9-qAwAJ).
+
+### Goals
+
+Delete the SSH Tunnel/Node Dialer code from Kube APIServer.
+Enable admins to fix https://groups.google.com/d/msg/kubernetes-security-announce/tyd-MVR-tY4/tyREP9-qAwAJ.
+Allow isolation of the Control network from the Cluster network.
+
+### Non-Goals
+
+Build a general purpose Proxy which does everything. (Users should build their own
+custom proxies with the desired behavior, based on the provided proxy)
+Handle requests from the Cluster to the Control Plane. (The proxy can be extended to
+do this. However that is left to the User if they want that behavior)
+
+## Definitions
+
+- **Master Network** An IP reachable network space which contains the master components, such as Kubernetes API Server,
+Connectivity Proxy and ETCD server.
+- **Node Network** An IP reachable network space which contains (some/all?) the clusters Nodes.
+Worth noting that the Node Network may be completely disjoint from the Master network.
+It may have overlapping IP addresses to the Master Network or other means of network isolation.
+Direct IP routability between cluster and master networks should not be assumed.
+- **KAS** Kubernetes API Server, responsible for serving the Kubernetes API to clients.
+- **KMS** Key Management Service, plugins for secrets encryption key management
+- **Connectivity service** A component built into the KAS which provides a golang dialer for outgoing connection requests.
+The dialer provided depends on NetworkContext information.
+- **Connectivity proxy** The proxy that runs in the master network.
+It has a secure channel established to the cluster network.
```

Comment:
following the same line of thought of nodes spanning multiple non-interconnected networks. it has a secure channel established to `cluster network`, or `cluster networks`?

Reply:
Please see above. In addition for cases like this I think we are likely to want to add some level of metadata on the request. (Eg. node name or failure domain/zone) So getting to the right network could either be by more fine grained ConnectivityServiceConfiguration or it would be that the proxy has enough metadata to make the call. Again though I think this is more part of beta than alpha.